core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/ForwardChainingSchemaCachingRDFSInferencer.java

```diff
@@ -0,0 +1,318 @@
+                }
+
+                // detect potential deadlock scenario
+                if(numberOfThreadsWaitingForWriteLock.get() > 1  //More than 1 connection waiting for a write lock
+                    && numberOfThreadsWaitingForWriteLock.get() <= readWriteLock.getReadLockCount() // no connection only using a read lock
+                    && random.nextBoolean()){ // randomly kick this connection out
+                    break;
+                }
+
+            }
+
+//            System.err.println("isReadLocked: " + readWriteLock.isReadLocked());
+//            System.err.println("isWriteLocked(): " + readWriteLock.isWriteLocked());
+//            System.err.println("read lock count: " + readWriteLock.getReadLockCount());
+
+            releaseLock(connection);
+            throw new SailConflictException("Concurrent modification of schema, could not acquire the lock.");
+        } finally {
+
+            numberOfThreadsWaitingForWriteLock.decrementAndGet();
+        }
+
+    }
+
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data) {
+        super(data);
+        schema = null;
+
+    }
+
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data, Repository schema) {
+        super(data);
+
+        this.schema = schema;
+
+    }
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data, boolean useAllRdfsRules) {
+        super(data);
+        schema = null;
+
+        this.useAllRdfsRules = useAllRdfsRules;
+
+    }
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data, Repository schema, boolean useAllRdfsRules) {
+        super(data);
+
+        this.schema = schema;
+        this.useAllRdfsRules = useAllRdfsRules;
+
+    }
+
+
+    public void initialize() throws SailException {
+        super.initialize();
+
+        if (sharedSchema) {
+            return;
+        }
+
+
+        ForwardChainingSchemaCachingRDFSInferencerConnection connection = null;
+
+        try {
+            connection = getConnection();
+            final ForwardChainingSchemaCachingRDFSInferencerConnection finalConnection = connection;
+            finalConnection.begin();
+
+            finalConnection.addAxiomStatements();
+
+
+            List<Statement> tboxStatments = new ArrayList<>();
+
+            if (schema != null) {
+
+                try (RepositoryConnection schemaConnection = schema.getConnection()) {
+                    schemaConnection.begin();
+                    RepositoryResult<Statement> statements = schemaConnection.getStatements(null, null, null);
+
+                    tboxStatments = Iterations.stream(statements)
+                        .peek(finalConnection::statementCollector)
+                        .collect(Collectors.toList());
+                    schemaConnection.commit();
+                }
+            }
+
+            finalConnection.calculateInferenceMaps();
+
+            if (schema != null) {
```

Comment:
This guard seems to be unnecessary, as if tboxStatements is empty, it will never iterate in forEach and seems to add another maintenance point in future if there is another condition other than "schema != null" added.